File: backend/src/services/db-operation.js
const sqlite3 = require('sqlite3').verbose();
const path = require('path');
const { DB_NAME, NODE_ENV } = require('../configs/constant');

const dbPath = NODE_ENV === 'development' ? path.resolve(__dirname, `../../${DB_NAME}`) : `/config/${DB_NAME}`;
const db = new sqlite3.Database(dbPath);

function initDb() {
  return new Promise((resolve, reject) => {
    db.run(`
      CREATE TABLE IF NOT EXISTS files (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        filename TEXT,
        original_path TEXT UNIQUE,
        original_size REAL,
        new_path TEXT,
        new_size REAL,
        created_at TEXT,
        processed_at TEXT,
        user_action TEXT,
        queue_status TEXT DEFAULT 'pending'
      )
    `, (err) => {
      if (err) reject(err);
      else resolve();
    });
  });
}

function addFile(filePath, name, size) {
  return new Promise((resolve, reject) => {
    db.run(
      `INSERT OR IGNORE INTO files (filename, original_path, original_size, created_at) VALUES (?, ?, ?, ?)`,
      [name, filePath, size, new Date().toISOString()],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

function getPendingFiles() {
  return new Promise((resolve, reject) => {
    db.all(
      `SELECT id, filename, original_path, original_size, created_at FROM files WHERE queue_status = 'pending'`,
      (err, rows) => {
        if (err) reject(err);
        else resolve(rows);
      }
    );
  });
}

function queueFiles(fileIds) {
  return new Promise((resolve, reject) => {
    const placeholders = fileIds.map(() => '(?)').join(',');
    db.run(
      `UPDATE files SET queue_status = 'confirmed' WHERE id IN (${placeholders})`,
      fileIds,
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

function getQueuedFiles() {
  return new Promise((resolve, reject) => {
    db.all(
      `SELECT id, filename, original_path, original_size, created_at FROM files WHERE queue_status = 'confirmed'`,
      (err, rows) => {
        if (err) reject(err);
        else resolve(rows);
      }
    );
  });
}


function getNextQueuedFile() {
  return new Promise((resolve, reject) => {
    db.get(
      `SELECT id, original_path FROM files WHERE queue_status = 'confirmed' LIMIT 1`,
      (err, row) => {
        if (err) reject(err);
        else resolve(row || null);
      }
    );
  });
}

function updateProcessing(fileId, newPath, newSize) {
  return new Promise((resolve, reject) => {
    db.run(
      `UPDATE files SET new_path = ?, new_size = ?, processed_at = ?, queue_status = 'processed' WHERE id = ?`,
      [newPath, newSize, new Date().toISOString(), fileId],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

function getProcessingFiles() {
  return new Promise((resolve, reject) => {
    db.all(
      `SELECT id, filename, original_path, original_size, new_size, created_at, processed_at FROM files WHERE queue_status = 'processed'`,
      (err, rows) => {
        if (err) reject(err);
        else resolve(rows);
      }
    );
  });
}

function updateReplaced(fileId, newPath, newSize) {
  return new Promise((resolve, reject) => {
    db.run(
      `UPDATE files SET original_path = ?, original_size = ?, new_path = NULL, new_size = NULL, user_action = 'replaced', queue_status = 'completed' WHERE id = ?`,
      [newPath, newSize, fileId],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

function updateDiscarded(fileId) {
  return new Promise((resolve, reject) => {
    db.run(
      `UPDATE files SET user_action = 'discarded', queue_status = 'completed' WHERE id = ?`,
      [fileId],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

function updateFailed(fileId) {
  return new Promise((resolve, reject) => {
    db.run(
      `UPDATE files SET queue_status = 'failed' WHERE id = ?`,
      [fileId],
      (err) => {
        if (err) reject(err);
        else resolve();
      }
    );
  });
}

module.exports = {
  initDb,
  addFile,
  getPendingFiles,
  queueFiles,
  getQueuedFiles,
  getNextQueuedFile,
  updateProcessing,
  getProcessingFiles,
  updateReplaced,
  updateDiscarded,
  updateFailed
};

File: backend/db_operation.py
from sqlite3 import connect
from datetime import datetime
from contextlib import contextmanager
from config import config

DB_PATH = config.DB_PATH

@contextmanager
def get_db_connection(db_path=DB_PATH):
    """Context manager for database connections"""
    conn = connect(db_path)
    try:
        yield conn
    finally:
        conn.close()

def init_db(db_path=DB_PATH):
    """Initialize the SQLite database and create the files table."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute('''CREATE TABLE IF NOT EXISTS files (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            original_path TEXT UNIQUE,
            original_size REAL,
            new_path TEXT,
            new_size REAL,
            created_at TEXT,
            processed_at TEXT,
            user_action TEXT,
            queue_status TEXT DEFAULT 'pending'
        )''')
        conn.commit()

def add_file(file_path, size, db_path=DB_PATH):
    """Add or update a file in the database."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("""
            INSERT OR IGNORE INTO files (original_path, original_size, created_at)
            VALUES (?, ?, ?)
        """, (file_path, size, datetime.now().isoformat()))
        conn.commit()

def get_pending_files(db_path=DB_PATH):
    """Retrieve all pending files from the database."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("SELECT id, original_path, original_size FROM files WHERE queue_status = 'pending'")
        return c.fetchall()

def queue_files(file_ids, db_path=DB_PATH):
    """Mark selected files as confirmed for processing."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.executemany("UPDATE files SET queue_status = 'confirmed' WHERE id = ?", [(id,) for id in file_ids])
        conn.commit()

def update_processing(file_id, new_path, new_size, db_path=DB_PATH):
    """Update file details after processing."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("""
            UPDATE files SET new_path = ?, new_size = ?, processed_at = ?, queue_status = 'processed'
            WHERE id = ?
        """, (new_path, new_size, datetime.now().isoformat(), file_id))
        conn.commit()

def update_replaced(file_id, new_path, new_size, db_path=DB_PATH):
    """Update file details after replacement."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("""
            UPDATE files SET original_path = ?, original_size = ?, new_path = NULL, new_size = NULL,
            user_action = 'replaced', queue_status = 'completed'
            WHERE id = ?
        """, (new_path, new_size, file_id))
        conn.commit()

def update_discarded(file_id, db_path=DB_PATH):
    """Mark file as discarded after processing."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("UPDATE files SET user_action = 'discarded', queue_status = 'completed' WHERE id = ?", (file_id,))
        conn.commit()

def update_failed(file_id, db_path=DB_PATH):
    """Mark file as failed if processing fails."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("UPDATE files SET queue_status = 'failed' WHERE id = ?", (file_id,))
        conn.commit()

def get_next_queued_file(db_path=DB_PATH):
    """Get the next file in the queue."""
    with get_db_connection(db_path) as conn:
        c = conn.cursor()
        c.execute("SELECT id, original_path FROM files WHERE queue_status = 'confirmed' LIMIT 1")
        return c.fetchone()
